src/v/kafka/server/handlers/produce.cc

```diff
@@ -190,6 +179,63 @@ static partition_produce_stages partition_append(
     };
 }
 
+static ss::future<model::record_batch> try_decompress(
+  bool is_compacted,
+  ss::foreign_ptr<std::unique_ptr<model::record_batch>> batch_ptr) {
+    auto batch = batch_ptr->copy();
+    if (is_compacted && batch.compressed()) {
+        return ss::do_with(std::move(batch), [](model::record_batch& batch) {
+            return storage::internal::decompress_batch(batch).then(
+              [&batch](model::record_batch&& b) {
+                  batch.decompressed(std::move(b));
```

Comment:
from what I can tell once we set the decompressed representation here it follows around the batch through copies, shares, etc... how does this affect memory usage in the batch cache? it looks like if we hold on to a batch with decompressed data in the cache we'll have 2 copies of the data?